src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -89,6 +91,83 @@ private void WriteObjectOrArrayStart(ClassType classType, JsonEncodedText proper
             }
         }
 
+        public void WritePreservedObjectOrArrayStart(ClassType classType, Utf8JsonWriter writer, JsonSerializerOptions options, string referenceId)
+        {
+            if (JsonPropertyInfo?.EscapedName.HasValue == true)
+            {
+                WritePreservedObjectOrArrayStart(classType, JsonPropertyInfo.EscapedName!.Value, writer, referenceId);
+            }
+            else if (KeyName != null)
+            {
+                JsonEncodedText propertyName = JsonEncodedText.Encode(KeyName, options.Encoder);
+                WritePreservedObjectOrArrayStart(classType, propertyName, writer, referenceId);
+            }
+            else
+            {
+                if ((classType & (ClassType.Object | ClassType.Dictionary)) != 0)
+                {
+                    writer.WriteStartObject();
+                    writer.WriteString("$id", referenceId);
+                    StartObjectWritten = true;
+                }
+                else
+                {
+                    // Wrap array into an object with $id and $values metadata properties.
+                    Debug.Assert(classType == ClassType.Enumerable);
+                    writer.WriteStartObject();
+                    writer.WriteString("$id", referenceId);
+                    writer.WritePropertyName("$values");
+                    writer.WriteStartArray();
```

Comment:
Combine the calls to WritePropertyName and WriteStartArray into one. But also use a `JsonEncodedText` instead.
```suggestion
                    writer.WriteStartArray("$values");
```